File: cpp/2075.decode-the-slanted-ciphertext.cpp
#include <string>
using std::string;

class Solution {
public:
  string decodeCiphertext(string encodedText, int rows) {
    const int n = encodedText.size();
    const int cols = n / rows;

    if (rows == 1) {
      return encodedText;
    }

    string res;

    for (int start = 0; start < cols; start++) {
      for (int i = 0, j = start; i < rows && j < cols; i++, j++) {
        res.push_back(encodedText[i * cols + j]);
      }
    }

    while (res.size() && res.back() == ' ') {
      res.pop_back();
    }

    return res;
  }
};

File: rust/2075.decode-the-slanted-ciphertext.rs
impl Solution {
    pub fn decode_ciphertext(encoded_text: String, rows: i32) -> String {
       if rows == 1 {
        return encoded_text;
       }

       let rows = rows as usize;
       let cols = encoded_text.len() / rows;
       let chars: Vec<char> = encoded_text.chars().collect();

       (0..cols)
        .flat_map(|start| {
          let chars = &chars;

          (0..rows)
            .take_while(move |&i| start + i < cols)
            .map(move |i| chars[i * cols + start + i])
        })
        .collect::<String>()
        .trim_end()
        .to_string()
    }
}